Preserve symlinks and empty directories in tar

mobileoverlord opened the pull request:
Empty directories and symlinks are currently omitted when packaging the tar file. This PR allows empty directories and symlinks to be retained. 

This should be rebased once https://github.com/hexpm/hex/pull/440 lands

mobileoverlord commented on lib/mix/tasks/hex.build.ex:
```diff
@@ -267,16 +267,16 @@ defmodule Mix.Tasks.Hex.Build do
     |> Enum.flat_map(&Path.wildcard/1)
     |> Enum.flat_map(&dir_files/1)
     |> Enum.map(&Path.expand/1)
-    |> Enum.filter(&File.regular?/1)
```
Removing this line makes me question what types of non-regular files we are interested in omitting.

ericmj replied:
I guess I just didn't see any reason for storing empty directories at the time.

ericmj commented on lib/hex/tar.ex:
```diff
@@ -38,9 +38,14 @@ defmodule Hex.Tar do
-          mode = File.stat!(name).mode
-          :ok = :hex_erl_tar.add(tar, contents, Hex.string_to_charlist(name), mode, [])
+          case File.lstat(name) do
+            {:ok, %File.Stat{type: type}} when type in [:directory, :symlink] ->
```
Won't this break for symlinks since we are not storing the target of the symlink? Maybe we should remove those, I don't know if `:erl_tar` supports symlinks?

ericmj replied:
Never mind, I just saw how you call `:erl_tar` below.

ericmj replied:
I believe that call below will add the current mtimes to the tar headers though which will break reproducibility. The reason we use `:hex_erl_tar.add/5` instead of `:hex_erl_tar.add/4` is so we don't preserve mtimes.

We don't seem to have tests that verifies this so can you please add assertions to the test that verifies we don't preserve mtimes for files or directories. If the test fails I will look into how we can solve this.

mobileoverlord commented:
I rebased this and added checks for mtime to the test. It started to fail now. If I understand correctly, the reason hex holds its own copy of erl_tar is to make the mtime modifications?

ericmj commented:
We vendor `erl_tar` to bring features from OTP 20 to all Hex users. Later we had to do modifications to support setting file modes while still setting mtime to unix epoch. I think we will have to do further modifications to support this use case, I will look into it today.

ericmj commented:
@mobileoverlord I pushed a few commits fixing the mtime issues for directories and symlinks and rebased your commits.

This is ready for review @wojtekmach.

wojtekmach commented on lib/hex/tar.ex:
```diff
@@ -38,9 +38,14 @@ defmodule Hex.Tar do
-          :ok = :hex_erl_tar.add(tar, contents, Hex.string_to_charlist(name), mode, [])
+          case Hex.file_lstat(name) do
+            {:ok, %File.Stat{type: type}} when type in [:directory, :symlink] ->
+              :ok == :hex_erl_tar.add(tar, Hex.string_to_charlist(name), [])
```
`:ok = :hex_erl_tar.add(...)`

wojtekmach commented on src/hex_erl_tar.erl:
```diff
@@ -1261,12 +1261,10 @@ convert_header(_Bin, _Reader) ->
-                             ctime=Fi#file_info.ctime,
+                             mtime=0,
+                             atime=0,
+                             ctime=0,
```
:+1:

btw, I noticed that it works well for files but not for dirs: (output from unpacked dummy package)

<img width="679" alt="zrzut ekranu 2017-10-22 o 19 49 31" src="https://user-images.githubusercontent.com/76071/31864701-564ad914-b762-11e7-8022-c533fa652a28.png">

I'm gonna create a separate issue for it if it's worth pursuing.

wojtekmach replied:
the output above is when I manually did: `tar xvzf contents.tar.gz`. I did manual unpacking because I thought there's a bug in `mix hex.build --smoke` (which I tried first) because when it unpacked it looked like it didn't preserve timestamps. However, while we're trying to preserve timestamps _in the archive_ when we unpack we're touching files [1] so I guess that was by design. Still I wonder if we should be zeroing timestamps for dirs in the archive.

[1] https://github.com/hexpm/hex/blob/master/lib/hex/tar.ex#L115

ericmj replied:
Yes, I have a comment about that in the tests. It would require bigger changes to erl_tar.

ericmj replied:
I'm not sure if we need to fix it for extracting unless someone actually asks for it, if `tar` has the same behavior then maybe it's even working as intended. What's important to ensure is that the tarballs we build do not include bad mtimes.

wojtekmach commented:
@milmazz just wanted to thank you again for your work on `mix hex.build --smoke` - it made testing this particular PR so much easier ❤️

milmazz commented:
@wojtekmach I'm glad you find it useful!
